sound/soc/sof/topology.c

```diff
@@ -641,18 +637,33 @@ static const struct sof_topology_token stream_tokens[] = {
 };
 
 /* Generic components */
-static const struct sof_topology_token comp_tokens[] = {
+const struct sof_topology_token comp_tokens[] = {
 	{SOF_TKN_COMP_PERIOD_SINK_COUNT,
-		SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
-		offsetof(struct sof_ipc_comp_config, periods_sink), 0},
+	 SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+	 offsetof(struct sof_ipc_comp_config, periods_sink), 0},
 	{SOF_TKN_COMP_PERIOD_SOURCE_COUNT,
-		SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
-		offsetof(struct sof_ipc_comp_config, periods_source), 0},
+	 SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+	 offsetof(struct sof_ipc_comp_config, periods_source), 0},
 	{SOF_TKN_COMP_FORMAT,
-		SND_SOC_TPLG_TUPLE_TYPE_STRING, get_token_comp_format,
-		offsetof(struct sof_ipc_comp_config, frame_fmt), 0},
+	 SND_SOC_TPLG_TUPLE_TYPE_STRING, get_token_comp_format,
+	 offsetof(struct sof_ipc_comp_config, frame_fmt), 0},
 };
 
+/* this should replace comp_tokens[] after switch all token parsing */
+const struct sof_topology_token comp_tokens_new[] = {
+	{SOF_TKN_COMP_PERIOD_SINK_COUNT,
+	 SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+	 offsetof(struct sof_ipc_comp_config, periods_sink), 0},
+	{SOF_TKN_COMP_PERIOD_SOURCE_COUNT,
+	 SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+	 offsetof(struct sof_ipc_comp_config, periods_source), 0},
+	{SOF_TKN_COMP_FORMAT,
+	 SND_SOC_TPLG_TUPLE_TYPE_STRING, get_token_comp_format_new,
+	 offsetof(struct sof_ipc_comp_config, frame_fmt), 0},
+};
```

Comment:
Let me see if I understand this right:
Currently we take the information from the topology and directly update a given ipc struct directly based on the sof_topology_token.

With this change we first take the topology information and convert it to an array of tuples.
Then we update the given ipc struct using the tuples array _and_ the sof_topology_token.
So if you want to support another type of ipc struct to update you would need another sof_topology_token struct to encode the offsets needed to be updated and also to have different sets of parameters?

Do we win anything apart from the intermediate step as I think we would be able to achieve this without introducing the tuples and just feed different sets of sof_topology_token to the topology parsing?

Reply:
the intermediate step is to speed things up. Parsing is a one time thing and can be done once at boot. If you were to parse the tokens and then set up the IPC struct in hw_params, we'd have latency concerns no?

Reply:
We only parsed at boot time, no? We set up the ipc struct once, no?
What I'm trying to say is that we can already do this w/o tuples:
```c
static const struct sof_topology_token ipc3_volume_tokens[] = {
	{SOF_TKN_VOLUME_RAMP_STEP_TYPE, SND_SOC_TPLG_TUPLE_TYPE_WORD,
		get_token_u32, offsetof(struct sof_ipc3_comp_volume, ramp), 0},
	{SOF_TKN_VOLUME_RAMP_STEP_MS,
		SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
		offsetof(struct sof_ipc3_comp_volume, initial_ramp), 0},
};

static const struct sof_topology_token ipc4_volume_tokens[] = {
	{SOF_TKN_VOLUME_RAMP_STEP_TYPE, SND_SOC_TPLG_TUPLE_TYPE_WORD,
		get_token_u32, offsetof(struct sof_ipc4_comp_volume, ramp), 0},
	{SOF_TKN_VOLUME_RAMP_STEP_MS,
		SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
		offsetof(struct sof_ipc4_comp_volume, initial_ramp), 0},
	{SOF_TKN_VOLUME_RAMP_STEP_MS,
		SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
		offsetof(struct sof_ipc4_comp_volume, mute_ramp), 0},
};

...
struct sof_ipc3_comp_volume *ipc3_volume;
struct sof_ipc4_comp_volume *ipc4_volume;

if (sof_is_ipc3(sdev))
	ret = sof_parse_tokens(scomp, ipc3_volume, ipc3_volume_tokens,
			ARRAY_SIZE(ipc3_volume_tokens), private->array,
			le32_to_cpu(private->size));
else
	ret = sof_parse_tokens(scomp, ipc4_volume, ipc4_volume_tokens,
			ARRAY_SIZE(ipc4_volume_tokens), private->array,
			le32_to_cpu(private->size));
```
Obviously not this exact way, but even if we have the tuples in between we need to have similar branching.
Atm we can do this way:
if (ipc3) convert the topology to ipc3 IPC struct
else convert the topology to ipc4 IPC struct

With the tuples in the middle:
convert the topology to tuples
if (ipc3) convert the tuples to ipc3 IPC struct
else convert the tuples to ipc4 IPC struct

Am I missing something?

Reply:
@ujfalusi I'm not sure I'm following you. Basically, if you make the parser IPC agnostic, it dont need to split up ipc3 tokens and ipc4 tokens. You could if you want to but you can still do without the if (sof_is_ipc3(sdev)).
```
ret = sof_parse_tokens(scomp, ipc3_volume, ipc3_volume_tokens,
			ARRAY_SIZE(ipc3_volume_tokens), private->array,
			le32_to_cpu(private->size));

ret = sof_parse_tokens(scomp, ipc4_volume, ipc4_volume_tokens,
			ARRAY_SIZE(ipc4_volume_tokens), private->array,
			le32_to_cpu(private->size));
```

Ideally a ipc3 based tplg would have ipc3 tokens only and ipc4 tplg would contain ipc4 tokens and then ipc struct builder would choose whatever tokens it need based on ipc version.

Reply:
@ujfalusi i've lost track of this conversation. So while the TOKEN IDs are common for different IPC versions, what token arrays they pertain to is IPC-specific. For ex: SOF_PCM_TOKENS could point to pcm_tokens[] in ipc3 and copier_tokens[] in ipc4.